polaris-service/src/main/java/org/apache/polaris/service/catalog/io/WasbTranslatingFileIOFactory.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.catalog.io;
+
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.io.FileIO;
+
+/** A {@link FileIOFactory} that translates WASB paths to ABFS ones */
+@JsonTypeName("wasb")
+public class WasbTranslatingFileIOFactory implements FileIOFactory {
+  @Override
+  public FileIO loadFileIO(String ioImpl, Map<String, String> properties) {
```

Comment:
At this point it's starting to seem that we need a way to do a chain of delegation. Both this class (or the suggested generalized `SchemeTranslatingFileIOFactory`) as well as `MeasuredFileIOFactory` fundamentally delegators, so maybe we should've entrenched an interface for delegation instead of just loading from `String ioImpl`.

If we make the delegation behavior configured as construction-time params or injection then we don't have to change the method signatures or callsites either. Basically like this:

    mainFileIOFactory = new SchemeTranslatingFileIOFactory(
        new MeasuredFileIOFactory(new DefaultFileIOFactory()),
        Map.of("wasb", "abfs", "wasbs", "abfss"));


And then we get rid of the hard-coded `CatalogUtil.loadFileIO` in both `SchemeTranslatingFileIOFactory` and `MeasuredFileIOFactory`:

    @Override
    public FileIO loadFileIO(String ioImpl, Map<String, String> properties) {
      return new SchemeTranslatingFileIO(this.delegateFactory.loadFileIO(ioImpl, properties);
    }

The drawback I guess is the config for setting the delegate factory is specific to each outer delegator factory:

    io:
      factoryType: scheme-translating
    schemeTranslatingDelegateType: measured
    measuredFactoryDelegateType: default

But this might be better than forcing delegation through a single config syntax since delegation might not be consistent for all delegator types. For example you might have something like:

    multiplexingIoFactoryDelegateTypes: r2,s3n,viewfs

Reply:
I can see how it could be a bigger undertaking to properly entrench how we want to convey such a delegation chain though, so I don't feel too strongly about whether to tackle this in this PR or in a later followup

Reply:
I don't think we can avoid working on this now. A production use case will require things like metering, throttling, as well as azure support. As is, the configuration only supports one IO factory, meaning you couldn't get all three without writing a new class that does all three.

I think the configuration would work fine as
```
io:
  factoryType: chain
    delegates:
      - measured
      - wasb
      - default
```
I think dropwizard should support detecting the generic type argument in something like
```
public void setDelegates(List<FileIOFactory> delegates);
```
so the above configuration should work out of the box.

Reply:
It won't work exactly out of the box, since FileIOFactory doesn't have a way to wrap a FileIO today. So it's not clear how the `chain` implementation would actually chain them. In [this feature branch](https://github.com/apache/polaris/compare/main...eric-maynard:polaris:delegating-io?expand=1) I toyed with adding a new interface called `SupportsFileIOWrapping` for this purpose, but I don't know that this is necessary just to get WASB working E2E